File: java/src/main/java/xyz/wasabicodes/jaws/struct/facet/Facet.java
package xyz.wasabicodes.jaws.struct.facet;

import xyz.wasabicodes.jaws.packet.Packet;
import xyz.wasabicodes.jaws.struct.User;

public interface Facet<C extends FacetContext<C>> {

    Class<? extends Facet<C>>[] getDependencies();

    void onInit(C ctx);

    void onStart(C ctx);

    void onEnd(C ctx);

    void onConnect(C ctx, User user);

    void onDisconnect(C ctx, User user);

    void onMessage(C ctx, User user, Packet received);

}

File: typescript/src/struct/facet.ts
import {User} from "./user";
import Packet from "../packet/packet";

export type FacetClass<T> = { new(): T, name: string };
export type FacetOrClass<T extends FacetContext<T>> = Facet<T> | FacetClass<Facet<T>>;

export interface FacetContext<C extends FacetContext<C>> {

    registerFacets(...facets: FacetOrClass<C>[]): void;

    getFacet<T extends Facet<C>>(clazz: FacetClass<T>): T | null;

    getFacetAssert<T extends Facet<C>>(clazz: FacetClass<T>): T;

    getAllFacets(): Facet<C>[];

}

export interface Facet<C extends FacetContext<C>> {

    onInit(ctx: C): void;

    onStart(ctx: C): void;

    onEnd(ctx: C): void;

    onConnect(ctx: C, user: User): void;

    onDisconnect(ctx: C, user: User): void;

    onMessage(ctx: C, user: User, received: Packet): void;

}
